### PR TITLE
Remove prophecy usage

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this is a changes on tests plus a patch on src

## Changelog

```markdown
### Fixed
- Call to deprecated `getTargetEntity`
```

This will allow us to have the php 8 build running without errors. Also it will allow to remove prophecy usage and to upgrade all builds to use phpunit 9 (Except php 7.2)
This is the first step, there are other changes needed on dev-kit and on other repositories

### CODE REVIEW

File: src/Form/DataTransformerResolver.php [hunk 1]
```diff
@@ -108,8 +108,8 @@ private function hasFieldDescriptionAssociationWithClass(
         FieldDescriptionInterface $fieldDescription,
         string $class
     ): bool {
-        return (
-            method_exists($fieldDescription, 'getTargetModel') && $class === $fieldDescription->getTargetModel()
-        ) || $class === $fieldDescription->getTargetEntity();
+        return method_exists($fieldDescription, 'getTargetModel') ?
```
Comment (jordisala1991) [inline src/Form/DataTransformerResolver.php:111]: I found this while changing all the tests, this can be done on a separate PR if you want.

**Comment by jordisala1991** [comment]
I spent a lot of time trying to fix the `BreadcrumbsBuilderTest`, this one `testUnitChildGetBreadCrumbs` I dont have idea on how to fix it.

```
There was 1 failure:

1) Sonata\AdminBundle\Tests\Admin\BreadcrumbsBuilderTest::testUnitChildGetBreadCrumbs
Failed asserting that actual size 0 matches expected size 5.

Error: /home/runner/work/SonataAdminBundle/SonataAdminBundle/tests/Admin/BreadcrumbsBuilderTest.php:403
```

@greg0ire maybe you can help me? I need to see what I broke with the change from prophecy to phpunit mocks.

File: src/Form/DataTransformerResolver.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Form/DataTransformerResolver.php:111]: This PR could take some times before being merged. So maybe it's indeed a good idea to create a separate PR

**Comment by jordisala1991** [comment]
I leave it as a draft, but the only thing missing before reviews is fixing that one test. The rest should be okay.

File: tests/Action/DashboardActionTest.php [hunk 2]
```diff
@@ -25,6 +25,11 @@
 
 class DashboardActionTest extends TestCase
 {
+    /**
+     * @var [type]
```
Comment (greg0ire) [inline tests/Action/DashboardActionTest.php:29]: What's this?

**Comment by greg0ire** [comment]
> @greg0ire maybe you can help me? I need to see what I broke with the change from prophecy to phpunit mocks.

Not sure what you broke, but when I wrote this test, I wasn't yet aware that mocking what you don't own is bad it seems. I pushed a new commit that greatly simplifies this test by relying on real classes instead of stubs.

File: tests/Bridge/Exporter/AdminExporterTest.php [hunk 3]
```diff
@@ -59,7 +59,7 @@ public function testGetExportFilename(): void
             ->method('getClass')
             ->willReturn('MyProject\AppBundle\Model\MyClass');
         $adminExporter = new AdminExporter(new Exporter());
-        $this->assertRegExp(
+        $this->assertMatchesRegularExpression(
```
Comment (greg0ire) [inline tests/Bridge/Exporter/AdminExporterTest.php:62]: Looks like you can't install a version of PHPUnit that has this method on 7.2. Maybe do this?
```suggestion
        $method = method_exists($this, 'assertMatchesRegularExpression')? 'assertMatchesRegularExpression' : 'assertRegExp';
        $this->$method(
```

**Comment by greg0ire** [comment]
My commit removed many expectations though, I will force push equivalent assertions soon.

File: tests/Bridge/Exporter/AdminExporterTest.php (hunk 3 above)
Comment (jordisala1991) [reply tests/Bridge/Exporter/AdminExporterTest.php:62]: I think I can fix it bumping phpunit-bridge: https://github.com/symfony/symfony/pull/37960

**Comment by jordisala1991** [comment]
> My commit removed many expectations though, I will force push equivalent assertions soon.

I pushed two more commits to fix builds and the var thing

File: tests/Action/DashboardActionTest.php (hunk 2 above)
Comment (jordisala1991) [reply tests/Action/DashboardActionTest.php]: Resolved!

File: tests/Bridge/Exporter/AdminExporterTest.php (hunk 3 above)
Comment (jordisala1991) [reply tests/Bridge/Exporter/AdminExporterTest.php:62]: Yes, fixed

**Comment by greg0ire** [comment]
I just force pushed (with lease)

**Comment by jordisala1991** [comment]
Aaaand the php8 is green again! Thank you @greg0ire .

Please review, @sonata-project/contributors

File: src/Form/DataTransformerResolver.php (hunk 1 above)
Comment (jordisala1991) [reply src/Form/DataTransformerResolver.php:111]: Let's see how this goes, if we see a lot of review process I will revert it from this PR and open a new one. If not, I will change it to patch and add a changelog

**Review by VincentLanglet — DISMISSED** [review]
Maybe some clean up is necessary in the commits ?

For instance the fix of `src/Form/DataTransformerResolver.php` should be in a separate commit.

File: src/Form/DataTransformerResolver.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Form/DataTransformerResolver.php:111]: Yes, but at least a separate commit seems needed.

**Comment by jordisala1991** [comment]
I squashed some commits, separate the code changed on `src`. It should be better now.

**Review by VincentLanglet — APPROVED** [review]

File: tests/Action/AppendFormFieldElementActionTest.php [hunk 4]
```diff
@@ -118,9 +110,9 @@ public function testAppendFormFieldElementAction(): void
 
     private function configureFormRenderer()
     {
-        $runtime = $this->prophesize(FormRenderer::class);
+        $runtime = $this->createStub(FormRenderer::class);
```
Comment (franmomu) [inline tests/Action/AppendFormFieldElementActionTest.php:113]:
```suggestion
        $runtime = $this->createMock(FormRenderer::class);
```
There is this call in `testAppendFormFieldElementAction`:
```php
$renderer->expects($this->once())->method('setTheme')->with($formView);
```

File: tests/Action/RetrieveAutocompleteItemsActionTest.php [hunk 5]
```diff
@@ -213,95 +233,100 @@ public function testRetrieveAutocompleteItemsComplexProperty(): void
...
-        $datagrid->setValue('_per_page', null, 10)->shouldBeCalled();
-        $datagrid->setValue('_page', null, 1)->shouldBeCalled();
-        $datagrid->buildPager()->willReturn(null);
-        $datagrid->getPager()->willReturn($pager->reveal());
-        $pager->getResults()->willReturn([$model]);
-        $pager->isLastPage()->willReturn(true);
-        $fieldDescription->getTargetModel()->willReturn(Foo::class);
-        $fieldDescription->getName()->willReturn('barField');
-        $fieldDescription->getAssociationAdmin()->willReturn($targetAdmin->reveal());
+        $targetAdmin = $this->createStub(AbstractAdmin::class);
```
Comment (franmomu) [inline tests/Action/RetrieveAutocompleteItemsActionTest.php:240]:
```suggestion
        $targetAdmin = $this->createMock(AbstractAdmin::class);
```

File: tests/Action/RetrieveFormFieldElementActionTest.php [hunk 6]
```diff
@@ -82,27 +81,27 @@ public function testRetrieveFormFieldElementAction(): void
             'context' => 'list',
         ], [], [], [], [], ['REQUEST_METHOD' => Request::METHOD_POST]);
 
-        $modelManager = $this->prophesize(ModelManagerInterface::class);
+        $modelManager = $this->createStub(ModelManagerInterface::class);
         $formView = new FormView();
-        $form = $this->prophesize(Form::class);
-        $formBuilder = $this->prophesize(FormBuilder::class);
+        $form = $this->createStub(Form::class);
```
Comment (franmomu) [inline tests/Action/RetrieveFormFieldElementActionTest.php:86]:
```suggestion
        $form = $this->createMock(Form::class);
```

File: tests/Admin/AdminTest.php [hunk 7]
```diff
@@ -2445,10 +2455,10 @@ public function testGetBreadcrumbs(): void
         $admin = $this->getMockForAbstractClass(AbstractAdmin::class, [
             'admin.my_code', 'My\Class', 'MyBundle\ClassAdminController',
         ]);
-        $builder = $this->prophesize(BreadcrumbsBuilderInterface::class);
+        $builder = $this->createStub(BreadcrumbsBuilderInterface::class);
```
Comment (franmomu) [inline tests/Admin/AdminTest.php:2458]:
```suggestion
        $builder = $this->createMock(BreadcrumbsBuilderInterface::class);
```

File: tests/Admin/AdminTest.php [hunk 8]
```diff
@@ -2459,13 +2469,12 @@ public function testBuildBreadcrumbs(): void
...
```
Comment (franmomu) [inline tests/Admin/AdminTest.php:2472]:
```suggestion
        $builder = $this->createMock(BreadcrumbsBuilderInterface::class);
```

File: tests/Admin/BreadcrumbsBuilderTest.php [hunk 9]
```diff
@@ -291,130 +292,107 @@ public function testGetBreadcrumbsWithNoCurrentAdmin(): void
...
         $breadcrumbsBuilder = new BreadcrumbsBuilder(['child_admin_route' => 'show']);
-        $admin = $this->prophesize(AbstractAdmin::class);
-        $admin->isChild()->willReturn(false);
-
-        $menuFactory = $this->prophesize(MenuFactory::class);
-        $menuFactory->createItem('root')->willReturn($menu);
-        $admin->getMenuFactory()->willReturn($menuFactory);
-        $labelTranslatorStrategy = $this->prophesize(
-            LabelTranslatorStrategyInterface::class
+        $admin = $this->createMock(AbstractAdmin::class);
```
Comment (franmomu) [inline tests/Admin/BreadcrumbsBuilderTest.php:297]:
```suggestion
        $admin = $this->createStub(AbstractAdmin::class);
```

File: tests/Controller/HelperControllerTest.php [hunk 10]
```diff
@@ -560,103 +604,114 @@ public function testRetrieveAutocompleteItemsComplexProperty(): void
...
-        $datagrid->setValue('_page', null, 1)->shouldBeCalled();
-        $datagrid->buildPager()->willReturn(null);
-        $datagrid->getPager()->willReturn($pager->reveal());
-        $pager->getResults()->willReturn([$model]);
-        $pager->isLastPage()->willReturn(true);
-        $fieldDescription->getTargetModel()->willReturn(Foo::class);
-        $fieldDescription->getName()->willReturn('barField');
-        $fieldDescription->getAssociationAdmin()->willReturn($targetAdmin->reveal());
+        $targetAdmin = $this->createStub(AbstractAdmin::class);
+        $datagrid = $this->createStub(DatagridInterface::class);
```
Comment (franmomu) [inline tests/Controller/HelperControllerTest.php:612]:
```suggestion
        $datagrid = $this->createMock(DatagridInterface::class);
```

File: tests/Controller/HelperControllerTest.php [hunk 11]
```diff
@@ -560,103 +604,114 @@ public function testRetrieveAutocompleteItemsComplexProperty(): void
...
+            ['req_param_name_page_number', null, '_page'],
+            ['to_string_callback', null, null],
+            ['target_admin_access_action', null, 'list'],
+        ]);
     }
 
     private function configureFormRenderer()
     {
-        $runtime = $this->prophesize(FormRenderer::class);
+        $runtime = $this->createStub(FormRenderer::class);
```
Comment (franmomu) [inline tests/Controller/HelperControllerTest.php:712]:
```suggestion
        $runtime = $this->createMock(FormRenderer::class);
```

File: tests/Filter/Persister/SessionFilterPersisterTest.php [hunk 12]
```diff
@@ -13,21 +13,21 @@
...
     /**
-     * @var SessionInterface|ObjectProphecy
+     * @var SessionInterface|MockObject
      */
     private $session;
 
     protected function setUp(): void
     {
-        $this->session = $this->prophesize(SessionInterface::class);
+        $this->session = $this->createStub(SessionInterface::class);
```
Comment (franmomu) [inline tests/Filter/Persister/SessionFilterPersisterTest.php:30]:
```suggestion
        $this->session = $this->createMock(SessionInterface::class);
```

File: tests/Twig/Extension/SonataAdminExtensionTest.php [hunk 13]
```diff
@@ -155,20 +154,24 @@ protected function setUp(): void
...
+        $this->templateRegistry = $this->createStub(TemplateRegistryInterface::class);
+        $this->container = $this->createStub(ContainerInterface::class);
+        $this->container->method('get')->with('sonata_admin_foo_service.template_registry')
+            ->willReturn($this->templateRegistry);
 
-        $this->securityChecker = $this->prophesize(AuthorizationCheckerInterface::class);
-        $this->securityChecker->isGranted(['foo', 'bar'], null)->willReturn(false);
-        $this->securityChecker->isGranted(Argument::type('string'), null)->willReturn(true);
+        $this->securityChecker = $this->createStub(AuthorizationCheckerInterface::class);
+        $this->securityChecker->method('isGranted')->willReturnMap([
```
Comment (franmomu) [inline tests/Twig/Extension/SonataAdminExtensionTest.php:163]: Please take a look at https://github.com/sonata-project/SonataAdminBundle/pull/6416#discussion_r495215330

File: tests/Util/ObjectAclManipulatorTest.php [hunk 14]
```diff
@@ -30,29 +29,29 @@ class ObjectAclManipulatorTest extends TestCase
 {
     protected function setUp(): void
     {
-        $this->output = $this->prophesize(OutputInterface::class);
-        $this->admin = $this->prophesize(AdminInterface::class);
+        $this->output = $this->createStub(OutputInterface::class);
```
Comment (franmomu) [inline tests/Util/ObjectAclManipulatorTest.php:32]:
```suggestion
        $this->output = $this->createMock(OutputInterface::class);
```

File: tests/Util/ObjectAclManipulatorTest.php [hunk 15]
```diff
@@ -30,29 +29,29 @@ class ObjectAclManipulatorTest extends TestCase
...
+        $this->admin = $this->createStub(AdminInterface::class);
```
Comment (franmomu) [inline tests/Util/ObjectAclManipulatorTest.php:33]:
```suggestion
        $this->admin = $this->createMock(AdminInterface::class);
```

**Review by franmomu — COMMENTED** [review]
I would replace:
```diff
- $this->container = $this->createStub(ContainerInterface::class);
+ $this->container = new Container();
```
But we could do that in another PR.

**Edit**: Another PR is fine 🤣

**Review by greg0ire — APPROVED** [review]

**Comment by greg0ire** [comment]
Oh sorry @franmomu I merged right when you published your review :sweat_smile:

**Comment by franmomu** [comment]
No worries, that could be change with the Container mock in another one.

**Comment by VincentLanglet** [comment]
@jordisala1991 Do you want to make a merge into master ? There should be some conflict.

**Comment by jordisala1991** [comment]
Yes, I will do it. I expect big conflict fight

**Comment by jordisala1991** [comment]
> No worries, that can be change with the Container mock in another one.

Can you make another PR to fix those things? I wont have time today to do it

**Pull request merged**